src/Donations/Migrations/RecalculateExchangeRate.php

```diff
@@ -0,0 +1,200 @@
+
+use Give\Donations\ValueObjects\DonationMetaKeys;
+use Give\Framework\Database\DB;
+use Give\Framework\Database\Exceptions\DatabaseQueryException;
+use Give\Framework\Migrations\Contracts\BatchMigration;
+use Give\Framework\Migrations\Exceptions\DatabaseMigrationException;
+use Give\Framework\QueryBuilder\QueryBuilder;
+use Give\Framework\Support\ValueObjects\Money;
+
+/**
+ * @unreleased
+ */
+class RecalculateExchangeRate extends BatchMigration
+{
+    /**
+     * @inheritDoc
+     */
+    public static function id(): string
+    {
+        return 'recalculate_exchange_rate';
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public static function title(): string
+    {
+        return 'Recalculate exchange rate';
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public static function timestamp(): string
+    {
+        return strtotime('2025-04-24 00:00:00');
+    }
+
+    /**
+     * Base query to find donations that need exchange rate recalculation
+     *
+     * @unreleased
+     */
+    protected function query(): QueryBuilder
+    {
+        // Select donations where base currency is not empty,
+        // base currency is different from payment currency,
+        // and either the exchange rate does not exist or is equal to 1
+
+        return DB::table('posts', 'posts')
+            ->leftJoin('give_donationmeta as baseCurrencyMeta', 'posts.ID', 'baseCurrencyMeta.donation_id')
+            ->leftJoin('give_donationmeta as paymentCurrencyMeta', 'posts.ID', 'paymentCurrencyMeta.donation_id')
+            ->leftJoin('give_donationmeta as exchangeRateMeta', 'posts.ID', 'exchangeRateMeta.donation_id')
+            ->where('posts.post_type', 'give_payment')
+            ->where('baseCurrencyMeta.meta_key', '_give_cs_base_currency')
+            ->whereIsNotNull('baseCurrencyMeta.meta_value')
+            ->where('paymentCurrencyMeta.meta_key', DonationMetaKeys::CURRENCY)
+            ->whereRaw('AND baseCurrencyMeta.meta_value != paymentCurrencyMeta.meta_value')
+            ->where('exchangeRateMeta.meta_key', DonationMetaKeys::EXCHANGE_RATE)
+            ->where(function($query) {
+                $query->where('exchangeRateMeta.meta_value', '1')
+                    ->orWhereIsNull('exchangeRateMeta.meta_value');
+            });
+    }
+
+    /**
+     * @inheritDoc
+     * @throws DatabaseMigrationException
+     */
+    public function runBatch($firstId, $lastId): void
+    {
+        try {
+            $query = $this->query()
+                ->select('posts.ID')
+                ->attachMeta(
+                    'give_donationmeta',
+                    'ID',
+                    'donation_id',
+                    [DonationMetaKeys::AMOUNT, 'amount']
+                )
+                ->attachMeta(
+                    'give_donationmeta',
+                    'ID',
+                    'donation_id',
+                    [DonationMetaKeys::CURRENCY, 'currency']
+                )
+                ->attachMeta(
```

Comment:
Check this https://github.com/impress-org/givewp/tree/develop/src/Framework/QueryBuilder#attachmeta

Reply:
https://github.com/impress-org/givewp/pull/7893/commits/4b173db6432fff34488ac3ec1e20ca04f73aa15a